### PR TITLE
Improve creation of jax-rs features

### PR DESCRIPTION
**Pull request by pavolloffay**
* Simplify registration of client filters by `client.register(clientFeafure)` client feature can be provided as a class or instance with custom configuration.
* Added `@Priority` to client filters.
* Simplified construction of server feature builder.

### CODE REVIEW

File: opentracing-jaxrs2/src/main/java/io/opentracing/contrib/jaxrs2/server/ServerTracingDynamicFeature.java [hunk 1]
```diff
@@ -21,16 +21,19 @@
...
+    /**
+     * When using this constructor application has to call {@link GlobalTracer#register} to register
+     * tracer instance. Ideally it should be called in {@link javax.servlet.ServletContextListener}.
+     */
     public ServerTracingDynamicFeature() {
         this(new Builder(GlobalTracer.get()));
     }
 
-    private ServerTracingDynamicFeature(Builder builder) {
+    public ServerTracingDynamicFeature(Builder builder) {
```
Comment (pavolloffay) [inline opentracing-jaxrs2/src/main/java/io/opentracing/contrib/jaxrs2/server/ServerTracingDynamicFeature.java]: API brainstorm:

Should be this private and only accessible by `Builder.build()`. I have made it public to be easier to find as the default constructor is public too.

@jpkrohling @objectiser thoughts?

Comment (pavolloffay) [inline opentracing-jaxrs2/src/main/java/io/opentracing/contrib/jaxrs2/server/ServerTracingDynamicFeature.java]: My concern is that if we keep it public users can generate grossness like this:

```java
register(new ServerTracingDynamicFeature(new Builder(tracer).withDecorator()));
```java

Comment (objectiser) [inline opentracing-jaxrs2/src/main/java/io/opentracing/contrib/jaxrs2/server/ServerTracingDynamicFeature.java]: I'd leave private for now until a case is made for it?

File: examples/spring-boot/src/main/java/io/opentracing/contrib/jaxrs2/example/spring/boot/JerseyConfig.java [hunk 2]
```diff
@@ -23,13 +22,9 @@
     @Inject
     public JerseyConfig(Tracer tracer) {
         Client client = ClientBuilder.newClient();
+        client.register(new ClientTracingFeature(new Builder(tracer)));
```
Comment (objectiser) [inline examples/spring-boot/src/main/java/io/opentracing/contrib/jaxrs2/example/spring/boot/JerseyConfig.java]: Can't this use the same pattern as the server, i.e. `client.register(new ClientTracingFeature.Builder(tracer).build())` - haven't checked the code so possibly this is supported.

Comment (pavolloffay) [inline examples/spring-boot/src/main/java/io/opentracing/contrib/jaxrs2/example/spring/boot/JerseyConfig.java]: You found it :+1: , see https://github.com/opentracing-contrib/java-jaxrs/pull/26#discussion_r115731035. I will make that constructor private.

File: opentracing-jaxrs2/src/main/java/io/opentracing/contrib/jaxrs2/server/ServerTracingDynamicFeature.java (hunk 1 above)
Comment (jpkrohling) [inline opentracing-jaxrs2/src/main/java/io/opentracing/contrib/jaxrs2/server/ServerTracingDynamicFeature.java]: What's wrong with `register(new ServerTracingDynamicFeature(new Builder(tracer).withDecorator()));` ?

File: opentracing-jaxrs2/src/main/java/io/opentracing/contrib/jaxrs2/client/ClientTracingFeature.java [hunk 3]
```diff
@@ -1,28 +1,45 @@
...
+import javax.ws.rs.core.Feature;
+import javax.ws.rs.core.FeatureContext;
+import javax.ws.rs.ext.Provider;
 
 /**
  * @author Pavol Loffay
  */
-public class ClientTracingFeature {
-
+@Provider
```
Comment (jpkrohling) [inline opentracing-jaxrs2/src/main/java/io/opentracing/contrib/jaxrs2/client/ClientTracingFeature.java]: This means that this is loaded up and installed automatically by the JAX-RS implementation, right?

Comment (pavolloffay) [inline opentracing-jaxrs2/src/main/java/io/opentracing/contrib/jaxrs2/client/ClientTracingFeature.java]: It does not work for clients, but I found using this annotation for client's features too. (maybe a good practice to stress that it is a provider)

File: opentracing-jaxrs2/src/main/java/io/opentracing/contrib/jaxrs2/server/ServerTracingDynamicFeature.java (hunk 1 above)
Comment (pavolloffay) [inline opentracing-jaxrs2/src/main/java/io/opentracing/contrib/jaxrs2/server/ServerTracingDynamicFeature.java]: It does not look so cool, my concern is that builder pattern is not used properly, but maybe it is a minor issue.

Comment (jpkrohling) [inline opentracing-jaxrs2/src/main/java/io/opentracing/contrib/jaxrs2/server/ServerTracingDynamicFeature.java]: You are designing an API. Every move you make should be well thought of: if you restrict access, you should know why. If you give access, you should know why :)

If you don't have a reason to give or restrict access, better to restrict now and give later, as it's harder to restrict it later if it's open today.

**Pull request merged**